frost-core/src/lib.rs

```diff
@@ -1,8 +1,96 @@
+/// pass-through, implemented for a type just for the ciphersuite, and calls through to another
+/// implementation underneath, so that this trait does not have to be implemented for types you
+/// don't own.
+pub trait Group {
+    /// An element of the scalar finite field that our group is defined over.
+    type Scalar;
+
+    /// An element of our group that we will be computing over.
+    type Element: Mul<Self::Scalar, Output = Self::Element> + Sub<Output = Self::Element>;
+
+    /// The order of the the quotient group when the prime order subgroup divides the order of the
+    /// full group.
+    ///
+    /// If using a prime order elliptic curve, the cofactor should be 1 in the scalar field.
+    const COFACTOR: Self::Scalar;
+
+    /// Additive [identity] of the prime order group.
+    ///
+    /// [identity]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-04.html#section-3.1-3.2
+    const IDENTITY: Self::Element;
+
+    /// The fixed generator element of the prime order group.
+    ///
+    /// The 'base' of [`ScalarBaseMult()`] from the spec.
+    /// [`ScalarBaseMult()`]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-04.html#section-3.1
+    const BASEPOINT: Self::Element;
+
+    /// Generate a random scalar from the entire space [0, l-1]
+    fn random_scalar<R: RngCore + CryptoRng>(rng: R) -> Self::Scalar;
+
+    /// Generate a random scalar from the entire space [1, l-1]
+    fn random_nonzero_scalar<R: RngCore + CryptoRng>(rng: R) -> Self::Scalar;
+}
+
+/// A [FROST ciphersuite] specifies the underlying prime-order group details and cryptographic hash
+/// function.
+///
+/// [FROST ciphersuite]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-04.html#name-ciphersuites
+pub trait Ciphersuite {
+    /// The prime order group (or subgroup) that this ciphersuite operates over.
+    type Group: Group;
+
+    /// H1 for a FROST ciphersuite.
+    ///
+    /// [spec]: https://github.com/cfrg/draft-irtf-cfrg-frost/blob/master/draft-irtf-cfrg-frost.md#cryptographic-hash
+    fn H1(m: &[u8]) -> &[u8];
+
+    /// H2 for a FROST ciphersuite.
+    ///
+    /// [spec]: https://github.com/cfrg/draft-irtf-cfrg-frost/blob/master/draft-irtf-cfrg-frost.md#cryptographic-hash
+    fn H2(m: &[u8]) -> &[u8];
+
+    /// H3 for a FROST ciphersuite.
+    ///
+    /// [spec]: https://github.com/cfrg/draft-irtf-cfrg-frost/blob/master/draft-irtf-cfrg-frost.md#cryptographic-hash
+    fn H3(m: &[u8]) -> &[u8];
+
+    /// Generates the challenge as is required for Schnorr signatures.
+    ///
+    /// Deals in bytes, so that [FROST] and singleton signing and verification can use it with different
+    /// types.
+    ///
+    /// This is the only invocation of the H2 hash function from the [RFC].
+    ///
+    /// [FROST]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-04.html#section-4.6
+    /// [RFC]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-04.html#section-3.2
+    fn challenge(
```

Comment:
Refactor in such a way that this can be shared nicely between singleton and FROST but less directly exposed? Maybe generic Schnorr methods

Reply:
Done as a standalone function made generic over an impl of `Ciphersuite`